### DISCUSSION TITLE
Project Overview for local workspace

### DISCUSSION THREAD

**Discussion by worksofliam** [codefori/vscode-ibmi#908] [Ideas]
I am thinking of some major enhancements to local development (as in RPGLE in your local workspace). In order to help the transition to git, we need even better tools for 'local' development.

### New view

I want to add a new view named 'Project Overview', which will add new functionality:

* Overview of all RPGLE service programs
   * Context item: Generate copybook for all exports
   * Shows all export procedures
      * Context item: Copy prototype to clipboard
      * Context item: show all references in entire project
      * Item command: Open up source code to definition line
* Overview of all RPGLE programs
   * Context item: show all references in entire project
   * Item command: Open up source code
* Overview of all files (tables/PFs and views/LFs mostly)
   * Context item: Show all references (which programs and service programs use this object)
   * Context item: Open up source code

### New editor functionality

* New Code Lens on export definitions:
   * Button: `${count} references` - count being how many references throughout the entire project. Button opens peek reference view
      * Possible example: https://cs.github.com/golang/vscode-go/blob/2e739cc05022296cab0cd7aa1919de948a595f21/src/goReferencesCodelens.ts#L92
   * Button: Copy prototype

* Feature to automatically create prototype
   * Depends on https://github.com/halcyon-tech/vscode-rpgle/pull/118
   * Adds all procedures to API list from entire project and can auto import them

### Outline view for `DCL-F` / `DCL-DS (EXTFILE)` will work on local sources

Instead of depending on the connection to get the columns, if it is a local source, see if we can find the file locally first. This does mean parsing local DDS and SQL.

## What needs work before this

* `vscode-rpgle` needs an integrated SQL parser and DDS parser
   * DDS parser can be stolen from `vscode-displayfile`
* For file definitions, RPGLE parser needs to look at local DDS/SQL.. somehow
* https://github.com/halcyon-tech/vscode-rpgle/pull/118 needs to get merged

## How project scanning will work

1. When `vscode-rpgle` loads, if a local workspace is detected, then it will scan the entire workspace and generate information for all RPGLE sources.
   * Including the linter with only references to collect those. They are required later
2. New file system listener, so when the file changes, we reload the information for that source

**Things to consider**

* Performance on large projects

**Comment by edmundreinhardt** [comment]
Will these projects expose the iproj.json project definition?

**Reply by edmundreinhardt** [reply]
It was intended to be a standard for IBMi git-based projects.  Feel free to suggest enhancements to it.

**Reply by worksofliam** [reply]
Yes, it will continue to support the use of `iproj.json` - though it won't need much info out of it other than the include directory information!

**Comment by edmundreinhardt** [comment]
I really like the create and copy prototype functions
Scott Klement has been asking for this for a while

**Comment by worksofliam** [comment]
Progress report!

### Showing all references to an export procedure

<img width="1596" alt="Screenshot 2022-10-17 at 11 14 54 PM" src="https://user-images.githubusercontent.com/3708366/196327299-6b0ea12c-45f2-4b0c-a0ac-dfc7aa0cfd25.png">

### Ability to open the implementation for a PR

![image](https://user-images.githubusercontent.com/3708366/196327233-c31175f0-c4ed-4800-9a03-2c7bd2a851d5.png)

**Comment by worksofliam** [comment]
PR created: https://github.com/halcyon-tech/vscode-rpgle/pull/140
